atc/scheduler/algorithm/group_resolver.go

```diff
@@ -256,6 +256,18 @@ outputs:
 				continue
 			}
 
+			if candidate == nil {
```

Comment:
What does it mean for `candidate` to be `nil` here? Does it mean there is currently no candidate version for a particular input?

If so, why don't we need to check for existence when `candidate != nil`? Could we not potentially overwrite the existing `candidate` with a version that doesn't exist?

Reply:
Looking at it more, I guess the `mismatch` case would apply if `candidate != nil` and the versions differed:

https://github.com/concourse/concourse/blob/e5f48f24ed9e7827547c1b71ee42b4df4523e15e/atc/scheduler/algorithm/group_resolver.go#L250-L253

Could you explain why we need to keep looking through builds if we already have a candidate version, and we don't even consider other versions? Or am I misunderstanding this part entirely?

Reply:
@aoldershaw For your first question, ya it is checking if there is no candidate set already for that version. If the version that we currently have is different than the one in the candidate, then that would be a mismatched version and we would restore that candidate to `nil`.

My original goal for having it only check if `candidate == nil` is so that when we vouch for that candidate with other jobs, we don't need to check if that version exists again. So the way that the algorithm works with passed constraints is that if you have one input with 3 passed constraints (let's say `job-1', `job-2` and `job-3`, it will recurse through the algorithm and "vouch for" the version we have determined to be the candidate with every passed constraint job. The flow is something like this
1. **candidate = nil, vouched for by = nil** : we run through the algorithm for the input and we run `trySatisfyPassedConstraintsForInput` for that input
https://github.com/concourse/concourse/blob/9f3c8970479647764fa1285e4574b08536eb1cbd/atc/scheduler/algorithm/group_resolver.go#L111-L116
then loop over the first passed constraint job `job-1`
https://github.com/concourse/concourse/blob/9f3c8970479647764fa1285e4574b08536eb1cbd/atc/scheduler/algorithm/group_resolver.go#L145
and determine a version as the candidate for that input (lets say the version is v1)
https://github.com/concourse/concourse/blob/9f3c8970479647764fa1285e4574b08536eb1cbd/atc/scheduler/algorithm/group_resolver.go#L270
then recurse
https://github.com/concourse/concourse/blob/9f3c8970479647764fa1285e4574b08536eb1cbd/atc/scheduler/algorithm/group_resolver.go#L282
1. **candidate = `v1`, vouched for by = `job-1`** : now we will `trySatisfyPassedConstraintsForInput` for that same input and this time when we loop over the passed constraint jobs, it will skip over `job-1` because it has already vouched for that version and it will move onto `job-2`
https://github.com/concourse/concourse/blob/9f3c8970479647764fa1285e4574b08536eb1cbd/atc/scheduler/algorithm/group_resolver.go#L148-L151
now `job-2` will then go through the same flow of finding a version for the candidate and if it finds the same version, then it will vouch for the version
https://github.com/concourse/concourse/blob/9f3c8970479647764fa1285e4574b08536eb1cbd/atc/scheduler/algorithm/group_resolver.go#L270
1. **candidate = `v1`, vouched for by = `job-1` and `job-2`** : the steps will be the same for `job-3` where it will skip over `job-1` and `job-2` and run `job-3`. If it produces `v1` then `job-3` will vouch for `v1`.
1. **candidate = `v1`, vouched for by = `job-1`, `job-2` and `job-3`** : we are finished! all the passed jobs have vouched for the version so now we can return out of this recursion. 
https://github.com/concourse/concourse/blob/9f3c8970479647764fa1285e4574b08536eb1cbd/atc/scheduler/algorithm/group_resolver.go#L180-L182

Based off of this, we will end up going over the code for vouching for a candidate version multiple times. I only wanted to check if the version exists if `candidate == nil` so that we only run the query for checking if the version exists once when we first determined a version for the candidate (which would be step 2 in the example above) rather if there was no check to see if candidate was nil then it would run the query every time we vouched for the candidate with all the passed constraint jobs.

Reply:
I think that comment might be a bit misleading?
```
// build contained a different version than the one we already have for 
// that candidate, so let's try a different build 
```

The `build` that it is referring to is a build of a different passed constraint job rather than the build of the passed constraint job that actually set the existing version on the candidate. So it would be when I have an input with two passed constraints, the first passed constraint job set the candidate to version `v1` but then when I recurse within the algorithm for the second passed constraint job, it finds `v2` as the version. This will end up being a mismatch and that will cause the `v1` version to be removed as a candidate.